myokit/formats/diffsl/_exporter.py

```diff
@@ -0,0 +1,523 @@
+#
+import collections
+import os
+import warnings
+
+import myokit
+import myokit.formats
+import myokit.formats.diffsl as diffsl
+import myokit.lib.guess as guess
+import myokit.lib.hh as hh
+import myokit.lib.markov as markov
+
+
+class DiffSLExporter(myokit.formats.Exporter):
+    """
+    This :class:`Exporter <myokit.formats.Exporter>` generates a DiffSL
+    implementation of a Myokit model.
+
+    Only the model definition is exported. All state variables and currents
+    are output, but no inputs are provided.
+
+    For details of the language, see https://martinjrobins.github.io/diffsl/
+    """
+
+    def model(self, path, model, protocol=None, convert_units=True):
+        """
+        Exports a :class:`myokit.Model` in DiffSL format, writing the result to
+        the file indicated by ``path``.
+
+        A :class:`myokit.ExportError` will be raised if any errors occur.
+
+        Arguments:
+
+        ``path``
+            The path to write the generated model to.
+        ``model``
+            The :class:`myokit.Model` to export.
+        ``protocol``
+            Not implemented!
+        ``convert_units``
+            If set to ``True``, the method will attempt to convert to
+            preferred units for voltage (mV), current (A/F), and time (ms).
+        """
+        # Raise exception if path is unwritable
+        self._test_writable_dir(os.path.dirname(path))
+
+        # Check model validity
+        try:
+            model.validate()
+        except myokit.MyokitError as e:
+            raise myokit.ExportError(
+                'DiffSL export requires a valid model.'
+            ) from e
+
+        # Prepare model for export
+        model, currents = self._prep_model(model, convert_units)
+
+        # Generate DiffSL model
+        diffsl_model = self._generate_diffsl(
+            model,
+            extra_out_vars=currents,
+        )
+
+        # Write DiffSL model to file
+        with open(path, 'w') as f:
+            f.write(diffsl_model)
+
+    def _prep_model(self, model, convert_units):
+        """
+        Prepare the model for export to DiffSL.
+        """
+        # Rewrite model so that any Markov models have a 1-sum(...) state
+        # This also clones the model, so that changes can be made
+        model = markov.convert_markov_models_to_compact_form(model)
+
+        # Remove all model bindings apart from time and pace
+        _ = myokit._prepare_bindings(
+            model,
+            {
+                'time': 't',
+                'pace': 'pace',
+            },
+        )
+
+        # Remove hardcoded stimulus protocol, if any
+        guess.remove_embedded_protocol(model)
+
+        # Get / try to guess some model variables
+        time = model.time()  # engine.time
+        vm = self._guess_membrane_potential(model)  # Vm
+        cm = guess.membrane_capacitance(model)  # Cm
+        currents = self._guess_currents(model)
+
```

Comment:
Should there be some check here that the model doesn't contain explicit time dependence?
I'm guessing that's not allowed in DiffSL

To implement, check if `time.refs_by()` is empty

Reply:
Resolved by 1a21249